### PR TITLE
circuit for oog log error state

### PR DESCRIPTION
**Pull request by DreamWuGit**
markdown spec  in spec repo https://github.com/privacy-scaling-explorations/zkevm-specs/pull/373. completed python spec after this merges!

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+        let gas_cost = GasCost::LOG.as_u64().expr()
+            + GasCost::LOG.as_u64().expr() * topic_count
+            + 8.expr() * memory_address.length()
+            + memory_expansion.gas_cost();
+
+        // Check if the amount of gas available is less than the amount of gas
+        // required
+        let insufficient_gas = LtGadget::construct(cb, cb.curr.state.gas_left.expr(), gas_cost);
+        cb.require_equal(
+            "log* gas left is less than gas required ",
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs:84]: log* -> logN

**Review by lispc — APPROVED** [review]

File: zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+        let mstart = cb.query_cell_phase2();
+        let msize = cb.query_word_rlc();
+        let rw_counter_end_of_reversion = cb.query_cell();
+
+        // Pop mstart_address, msize from stack
+        cb.stack_pop(mstart.expr());
+        cb.stack_pop(msize.expr());
+
+        // Note: no need to check not in static call, since write protection error will
+        // handle it.
```
Comment (ed255) [inline zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs:65]: Could you give more details on this?
Imagine we have the following scenario:
- gas_left < gas_log
- static = true
- opcode is LOG0

In that case I guess the error should be `ErrorWriteProtection` right?
But what forbids the prover to witness a `ErrorOOGLogGadget` instead?

File: zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+        );
+
+        // current call must be failed.
+        cb.call_context_lookup(false.expr(), None, CallContextFieldTag::IsSuccess, 0.expr());
+
+        cb.call_context_lookup(
+            false.expr(),
+            None,
+            CallContextFieldTag::RwCounterEndOfReversion,
+            rw_counter_end_of_reversion.expr(),
```
Comment (ed255) [inline zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs:96]: The `rw_counter_end_of_reversion` value is used in a read lookup, but there's no constraint associated to it.  Why is it read then?  I see that other errors do this:
```
rw_counter_end_of_step = cb.curr.state.rw_counter.expr() + cb.rw_counter_offset() - 1
require rw_counter_end_of_reversion == rw_counter_end_of_step + reversible_counter
```

**Review by ed255 — CHANGES_REQUESTED** [review]

File: zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs (hunk 2 above)
Comment (CPerezz) [reply zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs:65]: I think this fails within the same category of errors as https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1136 where we mistakenly assume that the prover will use the same code we provide. When indeed, nothing prevents a malicious party to modify the code. 

And so, either we have a constraint that handles correctly the `non_in_static_call` condition, or the Vk will never have such detail into account. And so as @ed255 said, we're still vulnerable.

File: zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+            "ErrorOutOfGasLOG happens in Log*",
+            opcode.expr(),
+            vec![
+                OpcodeId::LOG0.expr(),
+                OpcodeId::LOG1.expr(),
+                OpcodeId::LOG2.expr(),
+                OpcodeId::LOG3.expr(),
+                OpcodeId::LOG4.expr(),
+            ],
+        );
```
Comment (CPerezz) [inline zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs:55]: Not sure, but we should check if it's more expensive to do this, or instead, below, use the:
```rust
   let topic_count = opcode.expr() - OpcodeId::LOG0.as_u8().expr();
```
And constraint topic_count to be `<3`. I think the current solution requires much less constraints. But in any case, just mentioning.

File: zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+
+        // check memory
+        let memory_address = MemoryAddressGadget::construct(cb, mstart, msize);
+
+        // Calculate the next memory size and the gas cost for this memory
+        // access
+        let memory_expansion = MemoryExpansionGadget::construct(cb, [memory_address.address()]);
+
+        let gas_cost = GasCost::LOG.as_u64().expr()
+            + GasCost::LOG.as_u64().expr() * topic_count
```
Comment (CPerezz) [inline zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs:76]: This ends up in `cost(LOG_N) * (N+1)`. Right?? Why don't we do:
```suggestion
        let gas_cost = GasCost::LOG.as_u64().expr() * topic_count
```

File: zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs (hunk 3 above)
Comment (CPerezz) [reply zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs:96]: Agree. I thought it had to do with checking if the reversion was happening for the entire TX and so we were at the last RW, but this is done below on a different fashion.

File: zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+        let restore_context = cb.condition(1.expr() - cb.curr.state.is_root.expr(), |cb| {
+            RestoreContextGadget::construct(
+                cb,
+                0.expr(),
+                0.expr(),
+                0.expr(),
+                0.expr(),
+                0.expr(),
+                0.expr(),
+            )
```
Comment (CPerezz) [inline zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs:128]: We could maybe create an associated fn for `RestoreContextGadget` like: `RestoreContextGadget::from_revert_ctx` which directly places all the 0's and has less "magic numbers" arround.

File: zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+            offset,
+            F::from(step.gas_left),
+            F::from(step.gas_cost),
+        )?;
+
+        self.rw_counter_end_of_reversion.assign(
+            region,
+            offset,
+            Value::known(F::from(call.rw_counter_end_of_reversion as u64)),
+        )?;
```
Comment (CPerezz) [inline zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs:178]: If we remove this from the constraints, we should remove it here too.

**Review by CPerezz — CHANGES_REQUESTED** [review]

File: zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+    restore_context: RestoreContextGadget<F>,
+}
+
+impl<F: Field> ExecutionGadget<F> for ErrorOOGLogGadget<F> {
+    const NAME: &'static str = "ErrorOutOfGasLOG";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::ErrorOutOfGasLOG;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
```
[the remark this answers is not in the document]
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs:43]: @ed255  good catch !!

File: zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs (hunk 2 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs:65]: for evm execution flow:  1. precheck if out of gas , if yes, stop.  if not run 2. log instruction, in log instruction, check static call, if not , throw `ErrorWriteProtection` with stop.
I agree with you guys , to be cautious, will still add static call checking here before finding way to guarantee that .

File: zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs (hunk 1 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs]: updated in [920c06b](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1116/commits/920c06b0a2e727c0fd3a3c38db96eed0530b2c01) thanks

File: zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs:44]: added in [920c06b](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1116/commits/920c06b0a2e727c0fd3a3c38db96eed0530b2c01) thanks

File: zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs (hunk 2 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs:72]: add static call check in [920c06b](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1116/commits/920c06b0a2e727c0fd3a3c38db96eed0530b2c01) thanks

File: bus-mapping/src/evm/opcodes/error_oog_log.rs [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        geth_steps: &[GethExecStep],
+    ) -> Result<Vec<ExecStep>, Error> {
+        let geth_step = &geth_steps[0];
+        let mut exec_step = state.new_step(geth_step)?;
+        let next_step = if geth_steps.len() > 1 {
+            Some(&geth_steps[1])
+        } else {
+            None
+        };
+        exec_step.error = state.get_step_err(geth_step, next_step).unwrap();
```
[the remark this answers is not in the document]
Comment (DreamWuGit) [reply bus-mapping/src/evm/opcodes/error_oog_log.rs]: updated in [920c06b](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1116/commits/920c06b0a2e727c0fd3a3c38db96eed0530b2c01)

File: zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs (hunk 3 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs:103]: it is missing , added in [920c06b](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1116/commits/920c06b0a2e727c0fd3a3c38db96eed0530b2c01) as others

File: zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs (hunk 6 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs:135]: good idea , better to do it with common error gadget refactoring in https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1059

File: zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs (hunk 7 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs:198]: add constraint in [920c06b](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1116/commits/920c06b0a2e727c0fd3a3c38db96eed0530b2c01)

File: zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs (hunk 5 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs:83]: no, if opcode is LOG0 , `topic_count` is zero, then `gas_cost` will be zero by `let gas_cost = GasCost::LOG.as_u64().expr() * topic_count`

File: bus-mapping/src/evm/opcodes/error_oog_log.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (DreamWuGit) [reply bus-mapping/src/evm/opcodes/error_oog_log.rs]: I realize original `unwrap` method can assert there is an real error happen associate with the step,

File: zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs (hunk 5 above)
Comment (CPerezz) [reply zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs:76]: Ohh that makes sense! Thanks!

**Review by CPerezz — APPROVED** [review]
LGTM!! Thanks for this work @DreamWuGit

**Review by ed255 — APPROVED** [review]
LGTM!

**Pull request merged**